### ISSUE TITLE
Undecorate exported file names in .dll if compiling for x86 WINAPI with MSYS2/MINGW

### ISSUE THREAD

**Issue by bas-oversteegen** [hunspell/hunspell#553]
The issues can be of three types (put an X between the brackets):

- [ ] Bug reports
- [ ] Change request or feature request
- [X ] Others, questions

# Other issues
## Used software
| key                               | value |
|-----------------------------------|-------|
| Operating System: |   Windows 10    |
| Hunspell version:                |   1.6.2    |
| libhunspell .dll version:                |   1.6.0    |
| Command line tool: |   MSYS2/MINGW 2016.10.25    |

After compiling hunspell for Windows x86 __stdcall (WINAPI) with MSYS2/MINGW, in libhunspell-1.6-0.dll the exported function names are decorated as 'Function name@number of bytes in argument list', e.g. 'Hunspell_add@8'. I would like to have the names undecorated just as happens for Windows __cdecl x86 and for x64 (both __cdecl and __stdcall), i.e. 'Hunspell_add'. I didn't succeed although I tried several coding tips from the following documentation:

- https://stackoverflow.com/questions/38710243/dll-using-stdcall-without-name-decoration-why-does-it-even-work
- https://stackoverflow.com/questions/538134/exporting-functions-from-a-dll-with-dllexport?rq=1
- http://www.ni.com/white-paper/3056/en/](http://www.ni.com/white-paper/3056/en/
- https://groups.google.com/forum/#!topic/powersoft.public.powerbuilder.powerscript/WP23945qbXI

I need the __stdcall (WINAPI) approach because I want to wrap hunspell in PowerBuilder, which only supports WINAPI export functions. Because I want to target both x86 and x64 architectures, identical function names are easier to wrap (I will only need one declaration per function). 

### (1) Install hunspell
**a.** Download latest hunspell version from https://github.com/hunspell/hunspell/releases, currently 'hunspell-1.6.2.zip'.
**b.** Unzip 'hunspell-1.6.2.zip' to 'some short path without spaces\hunspell-1.6.2', e.g 'C:\hunspell-1.6.2'.
**c.** In '.\src\hunspell\hunspell.h', change first line of declaration for all export functions from

  LIBHUNSPELL_DLL_EXPORTED int Hunspell_add(Hunhandle* pHunspell,
  LIBHUNSPELL_DLL_EXPORTED int Hunspell_add_dic(Hunhandle* pHunspell,
  LIBHUNSPELL_DLL_EXPORTED int Hunspell_add_with_affix(Hunhandle* pHunspell,
  LIBHUNSPELL_DLL_EXPORTED int Hunspell_analyze(Hunhandle* pHunspell,
  LIBHUNSPELL_DLL_EXPORTED Hunhandle* Hunspell_create(const char* affpath,
  LIBHUNSPELL_DLL_EXPORTED Hunhandle* Hunspell_create_key(const char* affpath,
  LIBHUNSPELL_DLL_EXPORTED void Hunspell_destroy(Hunhandle* pHunspell);
  LIBHUNSPELL_DLL_EXPORTED void Hunspell_free_list(Hunhandle* pHunspell,
  LIBHUNSPELL_DLL_EXPORTED int Hunspell_generate(Hunhandle* pHunspell,
  LIBHUNSPELL_DLL_EXPORTED int Hunspell_generate2(Hunhandle* pHunspell,
  LIBHUNSPELL_DLL_EXPORTED char* Hunspell_get_dic_encoding(Hunhandle* pHunspell);
  LIBHUNSPELL_DLL_EXPORTED int Hunspell_remove(Hunhandle* pHunspell,
  LIBHUNSPELL_DLL_EXPORTED int Hunspell_spell(Hunhandle* pHunspell, const char*);
  LIBHUNSPELL_DLL_EXPORTED int Hunspell_stem(Hunhandle* pHunspell,
  LIBHUNSPELL_DLL_EXPORTED int Hunspell_stem2(Hunhandle* pHunspell,
  LIBHUNSPELL_DLL_EXPORTED int Hunspell_suggest(Hunhandle* pHunspell,

  to

  LIBHUNSPELL_DLL_EXPORTED int __stdcall Hunspell_add(Hunhandle* pHunspell,
  LIBHUNSPELL_DLL_EXPORTED int __stdcall Hunspell_add_dic(Hunhandle* pHunspell,
  LIBHUNSPELL_DLL_EXPORTED int __stdcall Hunspell_add_with_affix(Hunhandle* pHunspell,
  LIBHUNSPELL_DLL_EXPORTED int __stdcall Hunspell_analyze(Hunhandle* pHunspell,
  LIBHUNSPELL_DLL_EXPORTED Hunhandle* __stdcall Hunspell_create(const char* affpath,
  LIBHUNSPELL_DLL_EXPORTED Hunhandle* __stdcall Hunspell_create_key(const char* affpath,
  LIBHUNSPELL_DLL_EXPORTED void __stdcall Hunspell_destroy(Hunhandle* pHunspell);
  LIBHUNSPELL_DLL_EXPORTED void __stdcall Hunspell_free_list(Hunhandle* pHunspell,
  LIBHUNSPELL_DLL_EXPORTED int __stdcall Hunspell_generate(Hunhandle* pHunspell,
  LIBHUNSPELL_DLL_EXPORTED int __stdcall Hunspell_generate2(Hunhandle* pHunspell,
  LIBHUNSPELL_DLL_EXPORTED char* __stdcall Hunspell_get_dic_encoding(Hunhandle* pHunspell);
  LIBHUNSPELL_DLL_EXPORTED int __stdcall Hunspell_remove(Hunhandle* pHunspell,
  LIBHUNSPELL_DLL_EXPORTED int __stdcall Hunspell_spell(Hunhandle* pHunspell, const char*);
  LIBHUNSPELL_DLL_EXPORTED int __stdcall Hunspell_stem(Hunhandle* pHunspell,
  LIBHUNSPELL_DLL_EXPORTED int __stdcall Hunspell_stem2(Hunhandle* pHunspell,
  LIBHUNSPELL_DLL_EXPORTED int __stdcall Hunspell_suggest(Hunhandle* pHunspell,

### (2) Install MSYS2/MINGW
**a.** Download latest MSYS2 64 bit installer from http://www.msys2.org/, currently 'msys2-x86_64-20161025.exe', and run it;
**b.** Install in default directory 'C:\msys64', uncheck option 'Run MSYS2 64bit now.'
**c.** Create desktop shortcuts to 'C:\msys64\msys2.exe', 'C:\msys64\mingw32.exe' and 'C:\msys64\mingw64.exe'.
**d.** Run desktop shortcut to 'C:\msys64\msys2.exe' as administrator. MSYS terminal opens. In MSYS terminal:
  - run command 'pacman -Syu'.
If it doesn't reach the next prompt successfully, kill the MSYS process in Windows Task manager and again run desktop shortcut to 'C:\msys64\msys2.exe'.
  - repeatedly run command 'pacman -Syu' until there is nothing left to do.
  - run command 'pacman -S base-devel'.
  - run command 'pacman -S mingw-w64-i686-toolchain mingw-w64-i686-libtool'.
  - run command 'pacman -S mingw-w64-x86_64-toolchain mingw-w64-x86_64-libtool'.
  - run command 'exit' to close terminal.

### (3) Compile
**a.** To compile to Windows 32 bit (x86) .exe and .dll files,
  - Run desktop shortcut to 'C:\msys64\mingw32.exe' as administrator. MINGW32 terminal opens.
  - Disable your antivirus software to avoid annoying interference.
  - In MINGW32 terminal, run following commands:
      cd c:
      cd hunspell-1.6.2
      autoreconf -vfi
      ./configure
      make
  - Enable your antivirus software again.

or

**b.** To compile to Windows 64 bit (x64) .exe and .dll files,
  - Run desktop shortcut to 'C:\msys64\mingw64.exe' as administrator. MINGW64 terminal opens.
  - Disable your antivirus software to avoid annoying interference.
  - In MINGW64 terminal, run following commands:
      cd c:
      cd hunspell-1.6.2
      autoreconf -vfi
      ./configure
      make
  - Enable your antivirus software again.
### (4) Result
'.\src\hunspell\.libs\libhunspell-1.6-0.dll' (Hunspell Dynamic Link Library)

To verify the target architecture (32 or 64 bit), open the .dll file in a text editor and have a look at the header:
  32 bit (x86): PE  L
  64 bit (x64): PE  d†

DLL Export Viewer (http://www.nirsoft.net/utils/dll_export_viewer.html) lists the export functions of libhunspell-1.6-0.dll as:

  Windows x86
  --------------
  Hunspell_add@8
  Hunspell_add_dic@8
  Hunspell_add_with_affix@12
  Hunspell_analyze@12
  Hunspell_create@8
  Hunspell_create_key@12
  Hunspell_destroy@4
  Hunspell_free_list@12
  Hunspell_generate@16
  Hunspell_generate2@20
  Hunspell_get_dic_encoding@4
  Hunspell_remove@8
  Hunspell_spell@8
  Hunspell_stem@12
  Hunspell_stem2@16
  Hunspell_suggest@12

  Windows x64
  --------------
  Hunspell_add
  Hunspell_add_dic
  Hunspell_add_with_affix
  Hunspell_analyze
  Hunspell_create
  Hunspell_create_key
  Hunspell_destroy
  Hunspell_free_list
  Hunspell_generate
  Hunspell_generate2
  Hunspell_get_dic_encoding
  Hunspell_remove
  Hunspell_spell
  Hunspell_stem
  Hunspell_stem2
  Hunspell_suggest

## Question
So using __stdcall (WINAPI) export functions, I want the x86 names to be the same as the x64 names, i.e. undecorated. I need to do something extra in step (**1c**), but what? Using a #pragma or a .def file perhaps, but how exactly?

Any help is appreciated.

**Comment by dimztimz** [comment]
Thank you for you contribution.

What you are reporting is normal behavior. Please read on MSDN aboud stdcall. I really dont know why you need to change the calling convention.

**Comment by dimztimz** [comment]
Rereding this i can suggwst this.

First try to use the latest master where we fixed a bug with dllexport when not using MSVC.

Try to link to that code compiled with mingw. Notice that you will have to put mingw c++ standard libraru dll in PATH.

alternatively try to compile the library with Visual studio and then try linking to that dll.

Both of the above do it with normal (not stdcall) functions.

If both fail and stdcall must be used well i really dont know look around on msdn. It is not issue on our side.

**Comment by bas-oversteegen** [comment]
Taking the msvc direction I solved the exported function name decoration/mangling issue by adding to the libhunspell project a libhunspell.def file with the following contents:

LIBRARY "libhunspell"
EXPORTS
  Hunspell_add @1
  Hunspell_add_dic @2
  Hunspell_add_with_affix @3
  Hunspell_analyze @4
  Hunspell_create @5
  Hunspell_create_key @6
  Hunspell_destroy @7
  Hunspell_free_list @8
  Hunspell_generate @9
  Hunspell_generate2 @10
  Hunspell_get_dic_encoding @11
  Hunspell_remove @12
  Hunspell_spell @13
  Hunspell_stem @14
  Hunspell_stem2 @15
  Hunspell_suggest @16
  
These functions can now be accessed in libhunspell.dll undecorated/unmangled. 

By setting project property Calling Convention to '__stdcall (/Gz)', there was no need to individually declare the export functions with '__stdcall'.

Apart from the function name issue I had some problems accessing these functions in PowerBuilder. To solve this I defined wrapper functions that return/take integer object handles instead of HunspellImpl pointers. For the slst string array parameters I preallocate memory in PowerBuilder. The native slsts are freed in the wrappers such that in PowerBuilder I don't need to call Hunspell_free_list(). Furthermore I added some basic error handling. Lastly I added the function names to the libhunspell.def file.

All exported functions work as they should, except for native (and therefore also wrapper) Hunspell_generate() and Hunspell_generate2(), which return empty slsts. I will open a separate question issue for this.
